### PR TITLE
Add custom resize ratio to X.L.BinarySpacePartition

### PR DESCRIPTION
**Pull request by exorcist365**
### Description

Include a description for your changes, including the motivation
behind them.

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I've considered how to best test these changes (property, unit,
        manually, ...) and concluded: That it works.

  - [x] I updated the `CHANGES.md` file

  ~~- [ ] I updated the `XMonad.Doc.Extending` file (if appropriate)~~

### CODE REVIEW

**Comment by exorcist365** [comment]
> At a first glance, this seems like a breaking change. Any way we could avoid that?

Probably, I just don't know how.

**Comment by exorcist365** [comment]
Thank you, I will take a look!

**Comment by exorcist365** [comment]
I will reopen this when I have a better idea of how to enable the custom resize without breaking it.

**Comment by TheMC47** [comment]
@exorcist365 we can figure this out together! @liskin mentioned `PatternSynonyms`, which we can use here. What we can do is rename the constructors you added, so we don't have conflicts. Maybe `ExpandTowardsWith`? I'm bad at naming things. And then we can create a pattern for the old `ExpandTowards`:
```haskell
{-# LANGUAGE PatternSynonyms #-}

pattern ExpandTowards d = ExpandTowardsWith d 0.5
```
and only handle `ExpandTowardsWith` in the messages. Is it clear?

**Comment by TheMC47** [comment]
`ExpandTowards` are just patterns, they're not really messages. You can export them as `pattern ExpandTowards`, or by adding them to the list of constructors ( `ResizeDirectional(.., ExpandTowardsLeft)`)
https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/pattern_synonyms.html#import-and-export-of-pattern-synonyms

**Comment by exorcist365** [comment]
@TheMC47 thank you for helping me, I will proceed to pushing and reopen the pr.

File: XMonad/Layout/BinarySpacePartition.hs [hunk 1]
```diff
@@ -740,9 +748,12 @@ instance LayoutClass BinarySpacePartition Window where
                               , fmap move          (fromMessage m)
                               , fmap splitShift    (fromMessage m)
                               ]
-          resize (ExpandTowards dir) = growNthTowards dir b
-          resize (ShrinkFrom dir) = shrinkNthFrom dir b
-          resize (MoveSplit dir) = autoSizeNth dir b
+          resize (ExpandTowards dir) = growNthTowards dir 0.05 b
+          resize (ShrinkFrom dir) = shrinkNthFrom dir 0.05 b
+          resize (MoveSplit dir) = autoSizeNth dir 0.05 b
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:753]: As far as I understand how `PatternSynonyms` work, these cases are redundant as they're covered by the more general constructors

File: XMonad/Layout/BinarySpacePartition.hs [hunk 2]
```diff
@@ -108,37 +109,48 @@ import Data.Ratio ((%))
...
-data ResizeDirectional = ExpandTowards Direction2D | ShrinkFrom Direction2D | MoveSplit Direction2D deriving Typeable
+-- | Message for resizing one of the cells in the BSP
+data ResizeDirectional =
+        ExpandTowardsBy Direction2D Rational
+      | ShrinkFromBy Direction2D Rational
+      | MoveSplitBy Direction2D Rational deriving Typeable
 instance Message ResizeDirectional
 
--- |Message for rotating a split (horizontal/vertical) in the BSP
+-- | Patterns to replicate the old messages
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:127]: This comment is just for `ExpandTowards`. Please add a comment to each pattern, and say what it's equivalent to (e.g. `ExpandTowards L` is equivalent to `ExpandTowardsBy L 0.05`

**Review by TheMC47 — CHANGES_REQUESTED** [review]
Awesome work :tada:  I just have some minor comments.
It would be also nice if you could update the usage section to include the new constructors (maybe use them instead and add a note that `ExpandTowards x` is equivalent to `ExpandTowardsBy x 0.05`, or something similar)

File: XMonad/Layout/BinarySpacePartition.hs [hunk 3]
```diff
@@ -100,6 +100,9 @@ import Data.Ratio ((%))
 -- > , ("M-S-C-j",        sendMessage $ SplitShift Prev)
 -- > , ("M-S-C-k",        sendMessage $ SplitShift Next)
 --
+-- Note that ExpandTowards x is equal to ExpandTowardsBy x 0.05,
+-- ShrinkFrom x is equal to ShrinkFromBy x 0.05 and MoveSplit x is equal to MoveSplitBy x 0.05
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:104]:
```suggestion
-- Note that @ExpandTowards x@, @ShrinkFrom x@, and @MoveSplit x@ are
-- the same as respectively @ExpandTowardsBy x 0.05@, @ShrinkFromBy x 0.05@
-- and @MoveSplitBy x 0.05@.
```

File: XMonad/Layout/BinarySpacePartition.hs [hunk 4]
```diff
@@ -124,11 +127,15 @@ data ResizeDirectional =
       | MoveSplitBy Direction2D Rational deriving Typeable
 instance Message ResizeDirectional
 
--- | Patterns to replicate the old messages
+-- | ExpandTowards x is now the equivalent of ExpandTowardsBy x 0.05
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:130]:
```suggestion
-- | @ExpandTowards x@ is now the equivalent of @ExpandTowardsBy x 0.05@
```

File: XMonad/Layout/BinarySpacePartition.hs [hunk 5]
```diff
@@ -124,11 +127,15 @@ data ResizeDirectional =
...
 pattern ExpandTowards :: Direction2D -> ResizeDirectional
 pattern ExpandTowards d = ExpandTowardsBy d 0.05
+
+-- | ShrinkFrom x is now the equivalent of ShrinkFromBy x 0.05
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:134]:
```suggestion
-- | @ShrinkFrom x@ is now the equivalent of @ShrinkFromBy x 0.05@
```

File: XMonad/Layout/BinarySpacePartition.hs [hunk 6]
```diff
@@ -124,11 +127,15 @@ data ResizeDirectional =
...
 pattern ShrinkFrom :: Direction2D -> ResizeDirectional
 pattern ShrinkFrom d = ShrinkFromBy d 0.05
+
+-- | MoveSplit x is now the equivalent of MoveSplitBy x 0.05
```
Comment (TheMC47) [inline XMonad/Layout/BinarySpacePartition.hs:138]:
```suggestion
-- | @MoveSplit x@ is now the equivalent of @MoveSplitBy x 0.05@
```

**Review by TheMC47 — APPROVED** [review]
LGTM

**Comment by TheMC47** [comment]
Can you please squash the commits?

**Comment by exorcist365** [comment]
> Can you please squash the commits?

Sure.

**Review by slotThe — APPROVED** [review]

**Review by TheMC47 — CHANGES_REQUESTED** [review]
One thing and this is ready to go: you need to rebase instead of merging master into your branch.  This is actually mentioned in the contributing instructions https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md#rebasing-and-squashing-commits:

> Under no circumstances should you ever merge the master branch into your feature branch. This makes it nearly impossible to review your changes and we will not accept your PR if you do this.

Once you take care of that, we'll merge this. Thank you for your contribution once again :tada:

**Comment by liskin** [comment]
I'd also recommend going through https://chris.beams.io/posts/git-commit/ one more time. The "If merged, this …" commits indicate the author tried to learn about writing good commit messages, but hasn't got it yet.

**Comment by TheMC47** [comment]
The commit message is `Squash commits.` Can you please change it? You can use `git commit --amend` to edit it. I'd also kindly ask you to read the guide that was posted by liskin

**Comment by exorcist365** [comment]
I forgot to readd the changes suggested by slotThe so I will just push this.

**Comment by liskin** [comment]
The commits needed a bit more love but I decided it'll be easier if I do the final polish and merge it.
Thanks for the contribution! :tada:

**Comment by exorcist365** [comment]
Alright, thanks!

**Pull request merged**